### PR TITLE
Random removal, greedy insert

### PR DESCRIPTION
**Pull request by EmielK**

### CODE REVIEW

File: heuristic/classes/Route.py [hunk 1]
```diff
@@ -51,3 +51,16 @@ def _plan_cost(self, problem: Problem) -> float:
...
         return cost
+
+    def remove_customer(self, customer: int):
+        """
+        Removes customer from route and items with the customer as destination
+        or origin from the loading plan.
+        """
+        self.customers.remove(customer)
+
+        for stacks in self.plan:
```
Comment (N-Wouda) [inline heuristic/classes/Route.py]: This is dog slow. It's much better to do the following:

1. Determine the current index of customer in self.customers, and name it `idx`. Then:
    - All stacks up to `idx` have `d<customer>`, the customer's delivery item.
    - All stacks after `idx` have `p<customer>`, the customer's pickup item.
2. For each stack up to `idx`, remove `Item(problem.demands[customer], DEPOT, customer)` using `find_stack` on `Stacks`.
3. For each stack after `idx`, remove `Item(problem.pickups[customer], customer, DEPOT)`, again using `find_stack`.

File: heuristic/destroy_operators/Random_removal.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    and removes them from the solution.
+    """
+    destroyed = current.copy()
+
+    number_to_remove = customers_to_remove()
+
+    for customer in random_state.choice(destroyed.problem.num_customers,
+                                        number_to_remove,
+                                        replace=False):
+
```
Comment (N-Wouda) [inline heuristic/destroy_operators/Random_removal.py]: No need for white space.

File: heuristic/functions/customers_to_remove.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from ..constants import DEGREE_OF_DESTRUCTION
+from ..classes import Problem
+from numpy.random import RandomState
+
+
+def customers_to_remove():
+    return int(Problem.num_customers * DEGREE_OF_DESTRUCTION)
```
Comment (N-Wouda) [inline heuristic/functions/customers_to_remove.py]: You will need to pass a problem instance in - `Problem` is a class, not an instance. The class doesn't know about `num_customers`.

**Review by N-Wouda — CHANGES_REQUESTED** [review]

File: heuristic/classes/Route.py (hunk 1 above)
Comment (N-Wouda) [reply heuristic/classes/Route.py]: `find_stack` returns the stack containing the given item, in time `O(num_stacks)`. For this stack you will need to remove the item (there is no code for this yet, I believe, but a remove function on `Stack` should suffice).

File: heuristic/repair_operators/Random_repair.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+from random import randrange
+
+
+def random_repair(old_solution: Solution) -> Solution:
+    """
+    Makes a copy of the current solution. Randomly selects customer from the
+    unassigned list and places them in a random location in a random route.
+    Returns the resulting solution.
+    """
+    solution = old_solution.copy()
```
Comment (N-Wouda) [inline heuristic/repair_operators/Random_repair.py]: Repair operators do not need to copy.

File: heuristic/repair_operators/Random_repair.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+    while solution.unassigned:
```
Comment (N-Wouda) [inline heuristic/repair_operators/Random_repair.py]:
```python
while len(solution.unassigned) != 0:
```

File: heuristic/classes/Route.py [hunk 6]
```diff
@@ -51,3 +52,22 @@ def _plan_cost(self, problem: Problem) -> float:
...
+        Removes customer from route and items with the customer as destination
+        or origin from the loading plan.
+        """
+        idx = self.customers.index(customer)
+        delivery_item = Item(Problem.demands[customer], DEPOT, customer)
+        pickup_item = Item(Problem.demands[customer], customer, DEPOT)
+
+        for stacks in self.plan[:(idx - 1)]:
+            stack = stacks.find_stack(delivery_item)
+            stack.stack.remove(delivery_item)
```
Comment (N-Wouda) [inline heuristic/classes/Route.py]: This is not sufficient to remove the item - it also needs to be removed from the set member. Write a `remove` wrapper function on `Stack`.

File: heuristic/destroy_operators/Random_removal.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+from numpy.random import RandomState
+
+
+def random_removal(current: Solution, random_state: RandomState) -> Solution:
+    """
+    Copies the old solution, randomly places q customers in the unassigned list
+    and removes them from the solution.
+    """
+    destroyed = current.copy()
+    number_to_remove = customers_to_remove(Solution.problem)
```
Comment (N-Wouda) [inline heuristic/destroy_operators/Random_removal.py]: This does not work

File: heuristic/functions/customers_to_remove.py (hunk 3 above)
Comment (N-Wouda) [reply heuristic/functions/customers_to_remove.py]: Use `problem`.

File: heuristic/destroy_operators/Random_removal.py (hunk 7 above)
Comment (N-Wouda) [reply heuristic/destroy_operators/Random_removal.py]: Use `current`.

Comment (N-Wouda) [reply heuristic/destroy_operators/Random_removal.py]: Or `destroyed`, both would work.

File: heuristic/classes/Route.py [hunk 8]
```diff
@@ -51,3 +52,22 @@ def _plan_cost(self, problem: Problem) -> float:
...
 
         return cost
+
+    def remove_customer(self, customer: int):
+        """
+        Removes customer from route and items with the customer as destination
+        or origin from the loading plan.
+        """
+        idx = self.customers.index(customer)
+        delivery_item = Item(Problem.demands[customer], DEPOT, customer)
```
Comment (N-Wouda) [inline heuristic/classes/Route.py]: This does not work.

File: heuristic/classes/Route.py [hunk 9]
```diff
@@ -51,3 +52,22 @@ def _plan_cost(self, problem: Problem) -> float:
...
+        pickup_item = Item(Problem.demands[customer], customer, DEPOT)
```
Comment (N-Wouda) [inline heuristic/classes/Route.py]: Same here.

File: heuristic/classes/Route.py [hunk 10]
```diff
@@ -51,3 +52,22 @@ def _plan_cost(self, problem: Problem) -> float:
...
+
+        for stacks in self.plan[:(idx - 1)]:
```
Comment (N-Wouda) [inline heuristic/classes/Route.py]: Be _very_ careful about the offsets here - I do not know them by heart, and you'll need to test this to find out if this is the correct cut-off point. I think it's too early, but again, I could be wrong.

Comment (EmielK) [reply heuristic/classes/Route.py]: In the graphs, at idx there was only a pickup present, which is why I thought this would be the cut-off point.

Comment (N-Wouda) [reply heuristic/classes/Route.py]: Alright then!

File: heuristic/repair_operators/greedy_insert.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    TODO.
+    """
+    assigned = [customer for customer in
+                list(range(current.problem.num_customers)) if customer
+                not in current.unassigned]
+
+    #  per unassigned probeer in te voegen per route op elke plek en bereken
+    #  kosten, en ook kosten voor nieuwe route. Selecteer laagste en plaats daar
+
+    while current.unassigned != 0:
```
Comment (N-Wouda) [inline heuristic/repair_operators/greedy_insert.py]:
```python
len
```

File: heuristic/repair_operators/greedy_insert.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+        customer = next(customer for customer in current.unassigned)
```
Comment (N-Wouda) [inline heuristic/repair_operators/greedy_insert.py]:
```python
customer = current.unassigned.pop()
```

File: heuristic/repair_operators/greedy_insert.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+        # Closest visitable node.
+        nearest = np.argmin(current.problem.distances[assigned, customer])
+        # Find route it is in and insert customer after closest other customer.
```
Comment (N-Wouda) [inline heuristic/repair_operators/greedy_insert.py]: So I guess the plan here is the following:

```
IF nearest is the depot THEN
  Make route for customer
ELSE IF customer can be assigned to route containing the nearest customer THEN
  Assign customer to this route and update the loading plans
ELSE
  Make route for customer
```

File: heuristic/classes/Route.py (hunk 10 above)
Comment (N-Wouda) [reply heuristic/classes/Route.py]: The first loading plan is always just after leaving the depot, so this should run up to `idx + 1`, not `idx - 1`). Similarly for the pickup item, which should be in the tour from `idx + 1` onwards.

Comment (N-Wouda) [reply heuristic/classes/Route.py]: OK I need to think about this a little more.

Comment (N-Wouda) [reply heuristic/classes/Route.py]: Yeah it's up to `idx + 1` for deliveries, and from `idx + 1` onwards for pickups. See e.g. [this build](https://travis-ci.com/N-Wouda/OR-Analysis/jobs/288094552) where I have updated the validator to consider this.

File: heuristic/classes/Stacks.py [hunk 14]
```diff
@@ -23,6 +25,10 @@ def __iter__(self):
     def __getitem__(self, idx: int):
         return self.stacks[idx]
 
+    def copy(self):
+        tmp = copy.deepcopy(self)
```
Comment (N-Wouda) [inline heuristic/classes/Stacks.py]: Or just ```return copy.deepcopy(self)```. Also types.

File: heuristic/repair_operators/greedy_insert.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+from heuristic.classes import Item, Route, Solution, Stacks, Stack
+from heuristic.constants import DEPOT
+
+
+def greedy_insert(current: Solution, rnd_state: RandomState) -> Solution:
+    """
+    TODO.
+    """
+    while len(current.unassigned) != 0:
```
Comment (N-Wouda) [inline heuristic/repair_operators/greedy_insert.py:18]: This entire method is _far too slow_. Can you estimate the run-time cost?

File: heuristic/repair_operators/greedy_insert.py [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+            route._set.add(customer)  # TODO this is not very nice
+
+            # insert handling plan at added customer
+            stack_after_customer = Stacks.copy(route.plan[idx_customer])
+
+            route.plan.insert(idx_customer + 1, stack_after_customer)
+
+            # insert for delivery in all stacks before customer
+            for plan in route.plan[:idx_customer + 1]:
+                plan.shortest_stack().push_rear(delivery)
```
Comment (N-Wouda) [inline heuristic/repair_operators/greedy_insert.py:56]: There is no guarantee the tour capacities are respected when you do this.

File: heuristic/repair_operators/greedy_insert.py [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+            # insert pickup for all stacks after customer
+            for plan in route.plan[idx_customer + 1:]:
+                plan.shortest_stack().push_rear(pickup)
```
Comment (N-Wouda) [inline heuristic/repair_operators/greedy_insert.py:60]: Same here.

File: heuristic/classes/Route.py [hunk 18]
```diff
@@ -56,3 +56,72 @@ def _plan_cost(self, problem: Problem) -> float:
             cost += Stacks.cost(customer, problem, before, after)
 
         return cost
+
+    def is_feasible(self, problem: Problem, customer: int) -> bool:
```
Comment (N-Wouda) [inline heuristic/classes/Route.py]: I'd probably call this ```can_insert```

Comment (N-Wouda) [reply heuristic/classes/Route.py]: Also make ```problem``` the last argument, not the first.

File: heuristic/classes/Route.py [hunk 19]
```diff
@@ -56,3 +56,72 @@ def _plan_cost(self, problem: Problem) -> float:
...
+        """
+        Checks if inserting a customer in this route is feasible.
+        """
+        delivery = Item(problem.demands[customer], DEPOT, customer)
```
Comment (N-Wouda) [inline heuristic/classes/Route.py]: ```d_volume```, a `float` of just the delivery volume. No need for an Item here. Same for pickup, `p_volume`

File: heuristic/classes/Route.py [hunk 20]
```diff
@@ -56,3 +56,72 @@ def _plan_cost(self, problem: Problem) -> float:
...
+        pickup = Item(problem.pickups[customer], customer, DEPOT)
+
+        if all(stacks.shortest_stack().volume() + delivery.volume <=
```
Comment (N-Wouda) [inline heuristic/classes/Route.py]: Difficult to read. Split it into can_deliver and can_pickup, two boolean variables, and return `can_deliver and can_pickup`.

File: heuristic/classes/Route.py [hunk 21]
```diff
@@ -56,3 +56,72 @@ def _plan_cost(self, problem: Problem) -> float:
...
+        full_route += 1
+
+        idx += 1
+
+        result = problem.distances[full_route[idx - 1], customer + 1] + \
+                 problem.distances[customer + 1, full_route[idx]]
+
+        return result
+
+    def opt_insert(self, problem: Problem, customer: int) -> int:
```
Comment (N-Wouda) [inline heuristic/classes/Route.py]: Can this not re-use `insert_cost()`?

File: heuristic/repair_operators/greedy_insert.py [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+            idx = route.opt_insert(current.problem, customer)
+            cost = route.insert_cost(current.problem, idx, customer)
+
+            if route.is_feasible(current.problem, idx):
+                heapq.heappush(insertion_heap, (cost, idx, idx_route))
+
+        # Costs new route
+        cost_new_route = ([current.problem.distances[DEPOT + 1, customer + 1] +
+                           current.problem.distances[customer + 1, DEPOT + 1]])
+        heapq.heappush(insertion_heap, (cost_new_route, 0, len(current.routes)))
```
Comment (N-Wouda) [inline heuristic/repair_operators/greedy_insert.py]: Why does this need to be on the heap? Can't you compare this to the routes on the heap to determine if it's worthwhile? That'd allow you to store routes on the heap, which is just that much easier.

File: heuristic/repair_operators/greedy_insert.py [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+
+from heuristic.classes import Item, Route, Solution, Stacks
+from heuristic.constants import DEPOT
+
+
+def greedy_insert(current: Solution, rnd_state: RandomState) -> Solution:
+    """
+    TODO.
+    """
+
```
Comment (N-Wouda) [inline heuristic/repair_operators/greedy_insert.py:17]: Remove whitespace

File: heuristic/repair_operators/greedy_insert.py [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+import heapq
+
+
+from heuristic.classes import Item, Route, Solution, Stacks
+from heuristic.constants import DEPOT
+
+
+def greedy_insert(current: Solution, rnd_state: RandomState) -> Solution:
+    """
+    TODO.
```
Comment (N-Wouda) [inline heuristic/repair_operators/greedy_insert.py]: And add a description. This is related to `sequential best insertion` in Hornstra et al. (2020).

Comment (N-Wouda) [reply heuristic/repair_operators/greedy_insert.py]: Including a run-time complexity!

**Review by N-Wouda — APPROVED** [review]

**Comment by N-Wouda** [comment]
Mostly OK, LGTM.

**Pull request merged**